File: csharp/AuthorParser.cs

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;

namespace Markov
{
  public class AuthorParser
  {
    public static AuthorParser ForAuthor(string name) => new AuthorParser { _dir = name };
    const string BadChars = ",./\\\"'?!:;[]{}}|()-_\n\t";
    private string _dir;
    private Dictionary<string, int> _ngrams = new();

    public Task Parse() => Task.Run(() =>
    {
      var author = _dir.Split('/').Last();
      var files = Directory.EnumerateFiles(_dir);
      foreach (var file in files)
      {
        ParseFile(file);
      }

      var max = new KeyValuePair<string, int>("", 0);
      foreach (var kv in _ngrams)
      {
        if (kv.Value > max.Value)
        {
          max = kv;
        }
      }

      Console.WriteLine($"Author {author}: \"{max.Key}\" with {max.Value} repetitions.");
    });

    void ParseFile(string file)
    {
      var lines = File.ReadLines(file);

      var ngram = new List<string>();

      foreach (var line in lines.Select(RemoveUnwantedChars))
      {
        foreach (var word in line.Split(' '))
        {
          if (IsTooShort(word))
          {
            continue;
          }

          ngram.Add(word);

          if (ngram.Count != 3)
          {
            continue;
          }

          Increment(string.Join(' ', ngram));
          ngram.RemoveAt(0);
        }
      }
    }

    private void Increment(string key)
    {
      var val = _ngrams.GetValueOrDefault(key, 0);
      _ngrams[key] = val + 1;
    }

    private static string RemoveUnwantedChars(string line) => new(line.Select(c => BadChars.Contains(c) ? ' ' : c).ToArray());
    private static bool IsTooShort(string word) => word.Length < 3;
  }
}

File: dart/bin/author_parser.dart
import 'dart:io';

class AuthorParser {
  final String _dir;
  final _ngrams = <String, int>{};
  static const badChars = r"""[,.\\/\"'?!:;\[\]\{\}|\(\)-_\n\t]+""";

  AuthorParser(this._dir);

  Future<void> parse() async {
    final author = _dir.split('/').last;
    final files = Directory(_dir).list();

    final futures = <Future<void>>[];

    await files.forEach((file) => futures.add(parseFile(file.path)));

    await Future.wait(futures);

    var key = "";
    var val = 0;

    _ngrams.forEach((k, v) {
      if (v > val) {
        key = k;
        val = v;
      }
    });

    stdout.writeln('Author $author: "$key" with $val repetitions.');
  }

  Future<void> parseFile(String path) async {
    final lines = await File(path).readAsLines();
    final ngram = <String>[];

    for (final line in lines.map(toCleanString)) {
      for (final word in line.split(' ')) {
        if (word.length < 3) {
          continue;
        }

        ngram.add(word);
        if (ngram.length != 3) {
          continue;
        }

        _ngrams.update(ngram.join(' '), (value) => value + 1,
            ifAbsent: () => 1);

        ngram.remove(ngram.first);
      }
    }
  }

  String toCleanString(String input) =>
      input.toLowerCase().replaceAll(RegExp(badChars), ' ');
}
